### PR TITLE
[Update] Updated multidiceloss

### PR DESCRIPTION
**Pull request by MidoriyaHero**
Hi @DatacollectorVN,

I have attached all documents needed in [this Ticket](https://app.clickup.com/t/86eu638zc)

# 🧠 Understanding the Critical Bug in Multi-Class Dice Loss: Spatial Masking vs. Class Exclusion

This document explains a **critical flaw** in an original implementation of `MultiClassDiceLoss` that uses **spatial masking** to exclude the background class. We'll walk through a **concrete 3×3 matrix example** to show how this bug distorts training and leads to incorrect gradients. Then we'll compare it with the **correct approach**: excluding the background *class*, not the background *pixels*.

---

## 🎯 Goal of Dice Loss in Semantic Segmentation

The **Dice coefficient** measures the overlap between predicted segmentation and ground truth:

$$
\text{Dice} = \frac{2 \times |P \cap G|}{|P| + |G|}
$$

Where:
- $P$ = predicted segmentation (probabilities)
- $G$ = ground truth mask
- Higher Dice → better overlap

In deep learning, we use **Dice loss = 1 - Dice**, so lower is better.

For **multi-class segmentation**, we compute Dice per class and average.

---

## 🔧 Setup: Example Data

Let’s use a small `3×3` image with:
- **2 classes**: `0` = background, `1` = foreground (e.g., tumor)
- Batch size = 1
- `include_background = False` → we want to **exclude background from the loss**

### 📌 Ground Truth (`gts`)
```python
[[0, 0, 0],
 [0, 1, 1],
 [0, 1, 0]]
```
- 6 background pixels (label `0`)
- 3 foreground pixels (label `1`)

### 📌 Model Logits (`gts_pred`)
Raw outputs before softmax:

**Class 0 (background):**
```python
[[-1.0, -2.0, -1.5],
 [-1.2,  2.0,  1.8],
 [-0.8,  2.2, -1.0]]
```

**Class 1 (foreground):**
```python
[[ 1.0,  2.0,  1.5],
 [ 1.2, -2.0, -1.8],
 [ 0.8, -2.2,  1.0]]
```

---

## 🔁 Step 1: Compute Softmax Probabilities (`probs`)

Apply `F.softmax(dim=1)` across classes.

### ✅ `probs` – Class 0 (background)
```python
[[0.12, 0.05, 0.09],
 [0.09, 0.98, 0.94],
 [0.14, 0.97, 0.16]]
```

### ✅ `probs` – Class 1 (foreground)
```python
[[0.88, 0.95, 0.91],
 [0.91, 0.02, 0.06],
 [0.86, 0.03, 0.84]]
```

✅ The model is confident on foreground pixels (low prob for class 1), but **overconfident on background pixels for class 1** (many ~0.9 values).  
➡️ This model is actually **not good** — it misclassifies most background as foreground.

---

## 🔁 Step 2: One-Hot Encode Ground Truth

```python
one_hot = F.one_hot(gts, num_classes=2).permute(0,3,1,2).float()
```

### `one_hot[:, 0, :, :]` – Background
```python
[[1, 1, 1],
 [1, 0, 0],
 [1, 0, 1]]
```

### `one_hot[:, 1, :, :]` – Foreground
```python
[[0, 0, 0],
 [0, 1, 1],
 [0, 1, 0]]
```

---

# ❌ Original Code: Flawed "Spatial Masking"

```python
if not self.include_background:
    valid_mask = (gts != 0)  # ← masks out ALL pixels where label == 0
else:
    valid_mask = gts  # ← BUG! gts is not a boolean mask!
```

Then:
```python
valid_mask = valid_mask.unsqueeze(1).float()
probs = probs * valid_mask
one_hot = one_hot * valid_mask
```

So only **3 pixels** (where `gts == 1`) are kept. All others are zeroed out.

### 🚫 Consequences:
- **6 background pixels are ignored completely**
- Even if the model predicts them **very wrongly**, it **won’t be penalized**
- The Dice is computed only over the **3 foreground pixels**

---

### 🔢 Dice Calculation (Original – Only 3 Pixels)

We compute Dice **only on the 3 foreground pixels**:

| Pixel | `probs[1]` | `one_hot[1]` |
|------|------------|--------------|
| (1,1) | 0.02       | 1            |
| (1,2) | 0.06       | 1            |
| (2,1) | 0.03       | 1            |

- `intersection = (0.02 + 0.06 + 0.03) = 0.11`
- `sum_probs = 0.02 + 0.06 + 0.03 = 0.11`
- `sum_one_hot = 3`
- `cardinality = 0.11 + 3 = 3.11`
- `dice = (2×0.11 + 1) / (3.11 + 1) = 1.22 / 4.11 ≈ 0.297`
- `loss = 1 - 0.297 = 0.703`

🟢 **Looks decent**, but it’s **misleading**.

> ❌ The model is **not being punished** for assigning **high foreground probability (0.84–0.95)** to **6 background pixels**.

---

# ✅ New Code: Correct "Class Channel Slicing"

```python
if not self.include_background:
    probs = probs[:, 1:, :, :]      # keep all pixels, remove class 0
    one_hot = one_hot[:, 1:, :, :]
```

👉 **No spatial masking.** All 9 pixels are used.  
👉 Only **class 1 (foreground)** contributes to the loss.

---

### 🔢 Dice Calculation (New – All 9 Pixels)

We now compute Dice for **class 1 across the entire image**.

| Pixel | `probs[1]` | `one_hot[1]` |
|------|------------|--------------|
| (0,0) | 0.88       | 0            |
| (0,1) | 0.95       | 0            |
| (0,2) | 0.91       | 0            |
| (1,0) | 0.91       | 0            |
| (1,1) | 0.02       | 1            |
| (1,2) | 0.06       | 1            |
| (2,0) | 0.86       | 0            |
| (2,1) | 0.03       | 1            |
| (2,2) | 0.84       | 0            |

- `intersection = 0.02×1 + 0.06×1 + 0.03×1 = 0.11`
- `sum_probs = 0.88 + 0.95 + 0.91 + 0.91 + 0.02 + 0.06 + 0.86 + 0.03 + 0.84 = 6.46`
- `sum_one_hot = 3`
- `cardinality = 6.46 + 3 = 9.46`
- `dice = (2×0.11 + 1) / (9.46 + 1) = 1.22 / 10.46 ≈ 0.1166`
- `loss = 1 - 0.1166 = 0.8834`

🔴 **Much higher loss** — and **rightfully so**.

> ✅ The model is **penalized** for predicting high foreground probability in background regions.

---

# 📊 Comparison Summary

| Feature | ❌ Original (Spatial Mask) | ✅ New (Class Slice) |
|-------|----------------------------|---------------------|
| Pixels used | Only 3 (label = 1) | All 9 pixels |
| Background pixels ignored? | ✅ Yes (even correct predictions) | ❌ No |
| Model punished for false positives? | ❌ No | ✅ Yes |
| Denominator size | Small → unstable Dice | Large → stable |
| Dice value | ~0.297 (optimistic) | ~0.117 (realistic) |
| Loss | 0.703 | 0.883 |
| **Interpretation** | "How good am I where GT=1?" | "True overlap across full image" |

---

# 🛑 Why the Original Approach is Dangerous

The original method:
- **Rewards bad behavior**: A model can predict `class 1` everywhere except foreground and still get low loss.
- **Ignores generalization**: No incentive to learn that background should **not** be classified as foreground.
- **Biases training**: Overfits to foreground, under-penalizes false positives.
- **Contains a bug**: `else: valid_mask = gts` uses label tensor as mask — always zeros out background pixels.

---

# ✅ Correct Principle

> 🟩 **"Excluding background" means: do not compute Dice for class 0.**  
> 🟥 **It does NOT mean: ignore all pixels where the label is 0.**

You should:
- ✅ Use **all spatial locations** in the computation
- ✅ Compute Dice **per class**
- ✅ Exclude only the **background class** from the final average
- ✅ Keep false positives in background **penalized**

---

# ✅ Conclusion

| Takeaway | Explanation |
|--------|-------------|
| 🚫 Spatial masking (removing background pixels) is **wrong** | It distorts Dice and hides model errors |
| ✅ Class slicing (removing background class) is **correct** | Measures true overlap, penalizes false positives |
| 🔍 New code is **superior** | More accurate, stable, and meaningful gradients |
| ⚠️ But ensure clean labels or add `ignore_index` | Otherwise, out-of-range labels will crash `F.one_hot` |

> 🔥 **Bottom line**: The new version fixes a **critical conceptual bug** and leads to **better-trained models** with **more reliable segmentation**.
Thank you!

### CODE REVIEW

**Review by DatacollectorVN — APPROVED** [review]
LGTM
